src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -63550,6 +63538,245 @@ public void LocalFunctions_Speculation()
+                        }
+
+                        void f2()
+                        {
+                            f1();
+                        }
+                    }
+                }
+                """;
+            CreateCompilation(source).VerifyDiagnostics(
+                // (14,13): warning CS8602: Dereference of a possibly null reference.
+                //             x.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(14, 13),
+                // (17,14): warning CS8321: The local function 'f2' is declared but never used
+                //         void f2()
+                Diagnostic(ErrorCode.WRN_UnreferencedLocalFunction, "f2").WithArguments("f2").WithLocation(17, 14));
+        }
+
+        [Fact]
+        public void LocalFunction_Unreachable()
+        {
+            var source = """
+                #nullable enable
+                class C
+                {
+                    void M()
+                    {
+                        var x = "a";
+
+                        void f1()
+                        {
+                            x.ToString();
+                        }
+                    }
+                }
+                """;
+            CreateCompilation(source).VerifyDiagnostics(
+                // (8,14): warning CS8321: The local function 'f1' is declared but never used
+                //         void f1()
+                Diagnostic(ErrorCode.WRN_UnreferencedLocalFunction, "f1").WithArguments("f1").WithLocation(8, 14));
+        }
+
+        [Fact]
+        public void LocalFunction_Unreachable_Worsening()
+        {
+            var source = """
+                #nullable enable
+                class C
+                {
+                    void M()
+                    {
+                        var x = "a";
+
+                        void f1()
+                        {
+                            x.ToString(); // 1
+                            x = null;
+                            x.ToString(); // 2
+                        }
+                    }
+                }
+                """;
+            CreateCompilation(source).VerifyDiagnostics(
+                // (8,14): warning CS8321: The local function 'f1' is declared but never used
+                //         void f1()
+                Diagnostic(ErrorCode.WRN_UnreferencedLocalFunction, "f1").WithArguments("f1").WithLocation(8, 14),
+                // (12,13): warning CS8602: Dereference of a possibly null reference.
+                //             x.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(12, 13));
+        }
+
+        [Fact]
+        public void LocalFunction_Reachable()
+        {
+            var source = """
+                #nullable enable
+                class C
+                {
+                    void M()
+                    {
+                        var x = "a";
```

Comment:
Consider variants of this test (and maybe some others) with `var x = (string?)null;`